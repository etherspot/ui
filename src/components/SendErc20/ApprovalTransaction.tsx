/* MIT License
 *
 * Copyright (c) 2024 Etherspot
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import { EtherspotApprovalTransaction } from '@etherspot/transaction-kit';
import { isEmpty } from 'lodash';

// Utils
import { erc20ValidationMessage } from '../../utils/validation';

// Types
import type { SendERC20TransactionProps } from '../../models/Transactions';

/**
 * @name ApprovalTransaction
 * @description: The following block is the first transaction in this batch
    of transactions, and instructs Etherspot to set a spending
    limit for the Smart Contract located at `receiverAddress`
    to be allowed to spend the token located at `tokenAddress`
    up to the amount specified in `value`.
 * @param {SendERC20TransactionProps} props - The props for the component
 * @returns {React.ReactElement} The rendered component.
 */
function ApprovalTransaction(props: SendERC20TransactionProps) {
  const { approveFirst } = props;

  if (!isEmpty(erc20ValidationMessage(props)) || !approveFirst) return null;

  return <EtherspotApprovalTransaction {...props} />;
}

export default ApprovalTransaction;
