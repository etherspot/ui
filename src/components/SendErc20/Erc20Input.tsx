/* MIT License
 *
 * Copyright (c) 2023 Etherspot
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { useState } from 'react';

// Utils
import { normalizeRawValue, parseBigNumber } from '../../utils/bigNumber';
import { numberInProgressRegex } from '../../utils/common';
import { erc20ValidationMessage } from '../../utils/validation';

// Types
import type { SendERC20InputProps, SendERC20TransactionProps } from '../../models/Transactions';

/**
 * @name Erc20Input
 * @description: Use of Erc20Input as token value input.
 * @param {SendERC20InputProps & SendERC20TransactionProps} props - The props for the component
 * @returns {React.ReactElement} The rendered component.
 */
function Erc20Input(props: SendERC20InputProps & SendERC20TransactionProps) {
  const { value, errorMessageClassName, onChangeValue, decimals, className, disableSendOnEnter, handleEnterPress } =
    props;

  const [errorMessage, setErrorMessage] = useState('');

  // Check entered value is number or not and convert into bignumber.
  const onChange = (event) => {
    const { value: newValue } = event.target;

    const normalizedValue = normalizeRawValue(newValue, decimals);
    if (!numberInProgressRegex.test(normalizedValue)) return;

    const parseValue = parseBigNumber(normalizedValue);
    if (parseValue?.isNaN()) return;

    // Check tokenAddress, receiverAddress and value is valid or not.
    setErrorMessage(erc20ValidationMessage({ ...props, value: normalizedValue }));

    onChangeValue && onChangeValue(normalizedValue);
  };

  const onKeyDown = () => {
    if (!disableSendOnEnter) handleEnterPress && handleEnterPress();
  };

  return (
    <>
      <input
        type="text"
        name="value"
        id="value"
        value={value}
        onChange={onChange}
        className={className}
        onKeyDown={onKeyDown}
        placeholder="0.00"
      />
      {errorMessage ? <p className={`sm:text-sm ${errorMessageClassName}`}>{errorMessage}</p> : ''}
    </>
  );
}

export default Erc20Input;
