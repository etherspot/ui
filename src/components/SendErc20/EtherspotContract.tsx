/* MIT License
 *
 * Copyright (c) 2023 Etherspot
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import { EtherspotContractTransaction } from '@etherspot/transaction-kit';
import { isEmpty } from 'lodash';
import { utils } from 'ethers';

// Utils
import { erc20ValidationMessage } from '../../utils/validation';
import { parseBigNumber } from '../../utils/bigNumber';

// Types
import type { SendERC20TransactionProps } from '../../models/Transactions';

// ABI
import { ERC20ABI } from '../../data/abis/ERC20';

/**
 * @name EtherspotContract
 * @description: The following <EtherspotContract />
    component will transfer [value] [token] from the built-in
    Etherspot Smart Wallet account to the receiverAddress.
 * @param {SendERC20TransactionProps} props - The props for the component
 * @returns {React.ReactElement} The rendered component.
 */
function EtherspotContract(props: SendERC20TransactionProps) {
  if (!isEmpty(erc20ValidationMessage(props)) || !parseBigNumber(props.value)?.gt('0.0')) return null;

  return (
    <EtherspotContractTransaction
      contractAddress={props.tokenAddress}
      abi={ERC20ABI}
      methodName={'transfer'}
      params={[props.receiverAddress, utils.parseEther(props.value)]}
    />
  );
}

export default EtherspotContract;
