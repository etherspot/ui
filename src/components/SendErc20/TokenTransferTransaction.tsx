/* MIT License
 *
 * Copyright (c) 2023 Etherspot
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import { EtherspotTokenTransferTransaction } from '@etherspot/transaction-kit';
import { isEmpty } from 'lodash';

// Utils
import { erc20ValidationMessage } from '../../utils/validation';

// Types
import type { SendERC20TransactionProps } from '../../models/Transactions';

/**
 * @name TokenTransferTransaction
 * @description: The following <EtherspotTokenTransferTransaction />
    component will transfer [value] [token] from the built-in
    Etherspot Smart Wallet account to the receiverAddress.
 * @param {SendERC20TransactionProps} props - The props for the component
 * @returns {React.ReactElement} The rendered component.
 */
function TokenTransferTransaction(props: SendERC20TransactionProps) {
  if (!isEmpty(erc20ValidationMessage(props))) return null;

  return <EtherspotTokenTransferTransaction {...props} />;
}

export default TokenTransferTransaction;
