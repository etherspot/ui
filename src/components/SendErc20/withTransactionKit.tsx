/* MIT License
 *
 * Copyright (c) 2023 Etherspot
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import { EtherspotBatch, EtherspotBatches, EtherspotTransactionKit } from '@etherspot/transaction-kit';

// Types
import { SendERC20Props } from '../../models/Transactions';

// Required Props for the SendErc20 component
/**
 * @typedef {Object} SendERC20Props - Props for the SendERC20Props component.
 * @property {string} tokenAddress - Address of ERC20 token.
 * @property {string} receiverAddress - The Ethereum address of the receiver.
 * @property {string} value - value of Erc20 token.
 * @property {string} onChangeValue - gets updated value from onChangeValue.
 * @property {number} chainId - The Ethereum chain ID.
 * @property {ethers.Wallet} provider - The Ethereum provider.
 * @property {number} decimals - decimals for ERC20 token(using for convert number to bignumber).
 * @property {React.ReactNode | React.RefObject<HTMLElement>} [triggerElement] - Trigger element for UI interactions.
 */

/**
 * @name SendErc20
 * @description Using SendErc20 component to we can send Erc20 token. 
    For sending token need tokenAddress, receiverAddress and value.
 * @param {SendERC20Props} props - The props for the component
 * @returns {React.ReactElement} The rendered component.
 */
const withTransactionKit = (WrappedComponent) => {
  const TransactionKit = (props: SendERC20Props) => {
    const { chainId, provider } = props;

    return (
      <EtherspotTransactionKit provider={provider} chainId={chainId}>
        <EtherspotBatches>
          <EtherspotBatch>
            <WrappedComponent {...props} />
          </EtherspotBatch>
        </EtherspotBatches>
      </EtherspotTransactionKit>
    );
  };
  return TransactionKit;
};

export default withTransactionKit;
