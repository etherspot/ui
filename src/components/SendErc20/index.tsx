/* MIT License
 *
 * Copyright (c) 2023 Etherspot
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { useState } from 'react';
import { isEmpty } from 'lodash';
import { useEtherspotTransactions } from '@etherspot/transaction-kit';

// Types
import type { SendERC20Props } from '../../models/Transactions';

// Utils
import { errorLog } from '../../utils/common';
import { erc20ValidationMessage } from '../../utils/validation';

// Local
import EtherspotContract from './EtherspotContract';
import withTransactionKit from './withTransactionKit';
import ApprovalTransaction from './ApprovalTransaction';
import Erc20Input from './Erc20Input';

// Required Props for the SendErc20 component
/**
 * @typedef {Object} SendERC20Props - Props for the SendERC20Props component.
 * @property {string} tokenAddress - Address of ERC20 token.
 * @property {string} receiverAddress - The Ethereum address of the receiver.
 * @property {string} value - value of Erc20 token.
 * @property {string} onChangeValue - gets updated value from onChangeValue.
 * @property {number} chainId - The Ethereum chain ID.
 * @property {ethers.Wallet} provider - The Ethereum provider.
 * @property {number} decimals - decimals for ERC20 token(using for convert number to bignumber).
 */

/**
 * @name SendErc20
 * @description Using SendErc20 component to we can send Erc20 token. 
    For sending token need tokenAddress, receiverAddress and value.
 * @param {SendERC20Props} props - The props for the component
 * @returns {React.ReactElement} The rendered component.
 */
const SendErc20 = (props: SendERC20Props) => {
  const {
    debug,
    onlyEstimate,
    onSent,
    onEstimated,
    handleEnterPress,
    onExecutionStatus,
    containerClassName,
    buttonContainerClassName,
    buttonClassName,
    buttonTitle,
  } = props;

  const { estimate, send } = useEtherspotTransactions();

  const [isEstimate, setIsEstimate] = useState(false);

  // if disableSendOnEnter is false then fired this fn.
  const onEnterPress = () => {
    estimateAndSend();
    handleEnterPress && handleEnterPress();
  };

  // Execution for estimation and send process.
  const estimateAndSend = async () => {
    // Return if the token address, receiver address or value not valid.
    if (!isEmpty(erc20ValidationMessage(props)) || isEstimate) return;

    setIsEstimate(true);
    onExecutionStatus && onExecutionStatus(true);
    /**
       * Any transaction that is intended to be sent to the blockchain must be estimated first. 
         The estimation function performs several checks including cost estimation and transaction validity. 
         This method must always be called before we send, otherwise the send method will return an error.
       */
    let estimatedResponse;
    try {
      estimatedResponse = await estimate();
      onEstimated && onEstimated(estimatedResponse);
    } catch (e) {
      errorLog(
        'Etherspot UI: Sorry there was an error whilst estimating your transaction. Please visit our documentation for more information: https://etherspot.fyi',
        e,
        debug,
      );
    }

    if (!isEmpty(estimatedResponse) && !onlyEstimate) {
      /**
         * The send function simply sends all the EtherspotBatches, 
           which contain all EtherspotBatch and EtherspotTransaction 
           components to the blockchain via the Etherspot platform. 
        */
      try {
        const response = await send();
        onSent && onSent(response);
      } catch (e) {
        errorLog(
          'Etherspot UI: Sorry there was an error whilst sending your transaction. Please visit our documentation for more information: https://etherspot.fyi',
          e,
          debug,
        );
      }
    }

    setIsEstimate(false);
    onExecutionStatus && onExecutionStatus(false);
  };

  return (
    <div className={containerClassName}>
      <ApprovalTransaction {...props} />

      <EtherspotContract {...props} />

      <Erc20Input {...props} handleEnterPress={onEnterPress} />

      <div className={buttonContainerClassName}>
        <button onClick={estimateAndSend} className={buttonClassName}>
          {buttonTitle ?? (onlyEstimate ? 'Estimate' : 'Send')}
        </button>
      </div>
    </div>
  );
};

export default withTransactionKit(SendErc20);
