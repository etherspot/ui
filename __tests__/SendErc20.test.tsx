/* MIT License
 *
 * Copyright (c) 2024 Etherspot
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React from 'react';
import { ethers } from 'ethers';
import { render, fireEvent, cleanup, getByRole, getByTestId } from '@testing-library/react';
import SendErc20 from '../src/components/SendErc20';

afterEach(cleanup);

const ethersProvider = new ethers.providers.JsonRpcProvider('http://localhost:3000', 'goerli'); // replace with your node's RPC URL
const provider = ethers.Wallet.createRandom().connect(ethersProvider);

test('Send ERC-20 token renders correctly', () => {
  const tree = render(
    <SendErc20
      disableSendOnEnter
      approveFirst
      chainId={1}
      provider={provider}
      value={'0.1'}
      onChangeValue={() => {}}
      decimals={9}
      tokenAddress={'0xE17f017475a709De58E976081eB916081ff4c9d5'}
      receiverAddress={'0x0493b9a21dE42546B2E3687Da683D0B7B6ec2180'}
      className="block rounded-md border-0 px-3.5 py-1.5 pl-7 pr-20 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
    />,
  );
  expect(tree).toMatchSnapshot();
});

test('input value updates correctly', () => {
  // Arrange
  const { getByPlaceholderText } = render(
    <SendErc20
      disableSendOnEnter
      approveFirst
      chainId={1}
      provider={provider}
      onChangeValue={() => {}}
      decimals={9}
      tokenAddress={'0xE17f017475a709De58E976081eB916081ff4c9d5'}
      receiverAddress={'0x0493b9a21dE42546B2E3687Da683D0B7B6ec2180'}
      className="block rounded-md border-0 px-3.5 py-1.5 pl-7 pr-20 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
    />,
  );

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const inputElement: any = getByPlaceholderText('0.00');
  fireEvent.change(inputElement, { target: { value: '0.123' } });

  expect(inputElement.value).toBe('0.123');
});
