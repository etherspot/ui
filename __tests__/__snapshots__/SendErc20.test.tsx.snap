// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send ERC-20 token renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <input
          class="block rounded-md border-0 px-3.5 py-1.5 pl-7 pr-20 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
          id="value"
          name="value"
          placeholder="0.00"
          type="text"
          value="0.1"
        />
        <p
          class="sm:text-sm undefined"
        />
        <div>
          <button>
            Send
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <input
        class="block rounded-md border-0 px-3.5 py-1.5 pl-7 pr-20 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
        id="value"
        name="value"
        placeholder="0.00"
        type="text"
        value="0.1"
      />
      <p
        class="sm:text-sm undefined"
      />
      <div>
        <button>
          Send
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
